pyomo/dae/flatten.py

```diff
@@ -234,47 +239,108 @@ def slice_component_along_sets(
+    b: _BlockData
+        Block whose components will be sliced
+
+    index_stack: list
+        Sets above ``b`` in the block hierachy, including on its parent
+        component, that have been sliced. This is necessary to return the
+        sets that have been sliced.
+
+    slice_: IndexedComponent_slice or _BlockData
+        Slice generated so far.  This function will yield extensions to
+        this slice at the current level of the block hierarchy.
+
+    sets: ComponentSet of Pyomo sets
+        Sets that will be sliced
+
+    ctype: Subclass of Component
+        Type of components to generate
+
+    index_map: ComponentMap
+        Map from (some of) the specified sets to a "representative index"
+        to use when descending into subblocks. While this map does not need
+        to contain every set in the sliced sets, it must not contain any
+        sets that will not be sliced.
 
-    `index_stack` is a list of indices "above" `b` in the
-    hierarchy. Note that `b` is a data object, so any index
-    of its parent component should be included in the stack.
-
-    `slice_` is the slice generated so far. Our goal here is to
-    yield extensions to `slice_` at this level of the hierarchy.
-
-    `sets` is a ComponentSet of Pyomo sets that should be sliced.
-
-    `ctype` is the type we are looking for.
+    Yields
+    ------
+    
+    Tuple of Sets and an IndexedComponent_slice or ComponentData
+        The sets indexing the returned component or slice. If the component
+        is indexed, an IndexedComponent_slice is returned. Otherwise, a
+        ComponentData is returned.
 
-    `index_map` is potentially a map from each set in `sets` to a 
-    "representative index" to use when descending into subblocks.
     """
     if type(slice_) is IndexedComponent_slice:
         context_slice = slice_.duplicate()
     else:
         context_slice = None
 
-    # Looks for components indexed by these sets immediately in our block
-    for c in b.component_objects(ctype, descend_into=False):
+    # If active argument is specified and does not match the block's
+    # active flag, we return immediately. This matches the behavior of
+    # component_objects. We only need this check as we may modify the
+    # active argument sent to component_objects if ctype is not an
+    # ActiveComponent type.
+    if active is not None and active != b.active:
+        return
+
+    # Define this class so we don't have to call issubclass again later.
+    check_active = issubclass(ctype, ActiveComponent) and (active != None)
+
+    # If active=False and ctype is not an ActiveComponent (e.g. it is Var)
+    # we will not generate any components. To prevent this, only pass the
+    # active argument if we are looking for active components.
+    c_active = active if check_active else None
+
+    # Looks for components indexed by specified sets immediately in our block.
+    for c in b.component_objects(ctype, descend_into=False, active=c_active):
```

Comment:
@jsiirola This potentially gets slightly simpler if one of the `active=False` fixes we talked about gets merged. If we remove the `active=True` default for variables, then we can just pass `component_objects` the caller's active argument rather than our computed `c_active`. However, we still need to check if we are dealing with an ActiveComponent so we know whether to check the slice for an active data object below.